go/host/enclave/guardian.go

```diff
@@ -0,0 +1,550 @@
+	state         *StateTracker // state machine that tracks our view of the enclave's state
+	enclaveClient common.Enclave
+
+	sl guardianServiceLocator
+	db *db.DB
+
+	submitDataLock sync.Mutex // we only submit one block, batch or transaction to enclave at a time
+
+	batchInterval  time.Duration
+	rollupInterval time.Duration
+
+	running atomic.Bool
+	logger  gethlog.Logger
+}
+
+func NewGuardian(
+	cfg *config.HostConfig,
+	hostData host.Identity,
+	serviceLocator guardianServiceLocator,
+	enclaveClient common.Enclave,
+	db *db.DB,
+	logger gethlog.Logger,
+) *Guardian {
+	return &Guardian{
+		hostData:       hostData,
+		state:          NewStateTracker(logger),
+		enclaveClient:  enclaveClient,
+		sl:             serviceLocator,
+		batchInterval:  cfg.BatchInterval,
+		rollupInterval: cfg.RollupInterval,
+		db:             db,
+		logger:         logger,
+	}
+}
+
+func (g *Guardian) Start() error {
+	g.running.Store(true)
+	go g.mainLoop()
+	if g.hostData.IsSequencer {
+		// if we are a sequencer then we need to start the periodic batch/rollup production
+		// Note: after HA work this will need additional check that we are the **active** sequencer enclave
+		go g.periodicBatchProduction()
+		go g.periodicRollupProduction()
+	}
+
+	// subscribe for L1 and P2P data
+	g.sl.P2P().SubscribeForTx(g)
+	g.sl.L1Repo().Subscribe(g)
+	g.sl.L2Repo().Subscribe(g)
+
+	// start streaming data from the enclave
+	go g.streamEnclaveData()
+
+	return nil
+}
+
+func (g *Guardian) Stop() error {
+	g.running.Store(false)
+
+	err := g.enclaveClient.Stop()
+	if err != nil {
+		g.logger.Warn("error stopping enclave", "err", err)
+	}
+
+	err = g.enclaveClient.StopClient()
+	if err != nil {
+		g.logger.Warn("error stopping enclave client", "err", err)
+	}
+
+	return nil
+}
+
+func (g *Guardian) HealthStatus() host.HealthStatus {
+	// todo (@matt) do proper health status based on enclave state
+	errMsg := ""
+	if !g.running.Load() {
+		errMsg = "not running"
+	}
+	return &host.BasicErrHealthStatus{ErrMsg: errMsg}
+}
+
+func (g *Guardian) GetEnclaveState() *StateTracker {
+	return g.state
+}
+
+// GetEnclaveClient returns the enclave client for use by other services
+// todo (@matt) avoid exposing client directly and return errors if enclave is not ready for requests
+func (g *Guardian) GetEnclaveClient() common.Enclave {
+	return g.enclaveClient
+}
+
+// HandleBlock is called by the L1 repository. The host is subscribed to receive new blockg.
+func (g *Guardian) HandleBlock(block *types.Block) {
+	g.logger.Debug("Received L1 block", log.BlockHashKey, block.Hash(), log.BlockHeightKey, block.Number())
+	// record the newest block we've seen
+	g.state.OnReceivedBlock(block.Hash())
+	if !g.state.InSyncWithL1() {
+		return // ignore blocks until we're up-to-date
+	}
+	err := g.submitL1Block(block, true)
+	if err != nil {
+		g.logger.Warn("error processing L1 block", log.ErrKey, err)
+	}
+}
+
+// HandleBatch is called by the L2 repository when a new batch arrives
```

Comment:
would be good to explain here how the sequencer behaves differenty from  normal validator

Reply:
yeah good point, this should only be called for validators, I'll add a check in the code so we log an error if it is called unexpectedly.